File: original/Strategies/EnergyBalanceComponent.cs

using System;
using System.Collections.Generic;
using System.Linq;
using System.Xml;     

using SiriusQualityEnergyBalance.DomainClass;
namespace SiriusQualityEnergyBalance.Strategies
{
    public class EnergyBalanceComponent 
    {
        public EnergyBalanceComponent()
        {
        }

        public double albedoCoefficient
        {
            get
            {
                 return _NetRadiation.albedoCoefficient; 
            }
            set
            {
                _NetRadiation.albedoCoefficient = value;
            }
        }
        public double stefanBoltzman
        {
            get
            {
                 return _NetRadiation.stefanBoltzman; 
            }
            set
            {
                _NetRadiation.stefanBoltzman = value;
            }
        }
        public double elevation
        {
            get
            {
                 return _NetRadiation.elevation; 
            }
            set
            {
                _NetRadiation.elevation = value;
            }
        }
        public double lambdaV
        {
            get
            {
                 return _NetRadiationEquivalentEvaporation.lambdaV; 
            }
            set
            {
                _NetRadiationEquivalentEvaporation.lambdaV = value;
                _Penman.lambdaV = value;
                _CanopyTemperature.lambdaV = value;
            }
        }
        public double psychrometricConstant
        {
            get
            {
                 return _PriestlyTaylor.psychrometricConstant; 
            }
            set
            {
                _PriestlyTaylor.psychrometricConstant = value;
                _Penman.psychrometricConstant = value;
            }
        }
        public double Alpha
        {
            get
            {
                 return _PriestlyTaylor.Alpha; 
            }
            set
            {
                _PriestlyTaylor.Alpha = value;
                _Penman.Alpha = value;
                _PtSoil.Alpha = value;
            }
        }
        public double vonKarman
        {
            get
            {
                 return _Conductance.vonKarman; 
            }
            set
            {
                _Conductance.vonKarman = value;
            }
        }
        public double heightWeatherMeasurements
        {
            get
            {
                 return _Conductance.heightWeatherMeasurements; 
            }
            set
            {
                _Conductance.heightWeatherMeasurements = value;
            }
        }
        public double zm
        {
            get
            {
                 return _Conductance.zm; 
            }
            set
            {
                _Conductance.zm = value;
            }
        }
        public double d
        {
            get
            {
                 return _Conductance.d; 
            }
            set
            {
                _Conductance.d = value;
            }
        }
        public double zh
        {
            get
            {
                 return _Conductance.zh; 
            }
            set
            {
                _Conductance.zh = value;
            }
        }
        public double soilDiffusionConstant
        {
            get
            {
                 return _DiffusionLimitedEvaporation.soilDiffusionConstant; 
            }
            set
            {
                _DiffusionLimitedEvaporation.soilDiffusionConstant = value;
            }
        }
        public double rhoDensityAir
        {
            get
            {
                 return _Penman.rhoDensityAir; 
            }
            set
            {
                _Penman.rhoDensityAir = value;
                _CanopyTemperature.rhoDensityAir = value;
            }
        }
        public double specificHeatCapacityAir
        {
            get
            {
                 return _Penman.specificHeatCapacityAir; 
            }
            set
            {
                _Penman.specificHeatCapacityAir = value;
                _CanopyTemperature.specificHeatCapacityAir = value;
            }
        }
        public double tauAlpha
        {
            get
            {
                 return _PtSoil.tauAlpha; 
            }
            set
            {
                _PtSoil.tauAlpha = value;
            }
        }
        public int isWindVpDefined
        {
            get
            {
                 return _EvapoTranspiration.isWindVpDefined; 
            }
            set
            {
                _EvapoTranspiration.isWindVpDefined = value;
            }
        }

        public void Estimate(SiriusQualityEnergyBalance.DomainClass.EnergyBalanceState s,SiriusQualityEnergyBalance.DomainClass.EnergyBalanceState s1,SiriusQualityEnergyBalance.DomainClass.EnergyBalanceRate r,SiriusQualityEnergyBalance.DomainClass.EnergyBalanceAuxiliary a,SiriusQualityEnergyBalance.DomainClass.EnergyBalanceExogenous ex)
        {
            try
            {
                CalculateModel(s, s1, r, a, ex);
            }
            catch (Exception exception)
            {
                string msg = "Error in component SiriusQualityEnergyBalance, strategy: " + this.GetType().Name + ": Unhandled exception running model. "+exception.GetType().FullName+" - "+exception.Message;
                throw new Exception(msg, exception);
            }
        }

        private void CalculateModel(SiriusQualityEnergyBalance.DomainClass.EnergyBalanceState s,SiriusQualityEnergyBalance.DomainClass.EnergyBalanceState s1,SiriusQualityEnergyBalance.DomainClass.EnergyBalanceRate r,SiriusQualityEnergyBalance.DomainClass.EnergyBalanceAuxiliary a,SiriusQualityEnergyBalance.DomainClass.EnergyBalanceExogenous ex)
        {
            EstimateOfAssociatedClasses(s, s1, r, a, ex);
        }

        //Declaration of the associated strategies
        NetRadiation _NetRadiation = new NetRadiation();
        NetRadiationEquivalentEvaporation _NetRadiationEquivalentEvaporation = new NetRadiationEquivalentEvaporation();
        PriestlyTaylor _PriestlyTaylor = new PriestlyTaylor();
        Conductance _Conductance = new Conductance();
        DiffusionLimitedEvaporation _DiffusionLimitedEvaporation = new DiffusionLimitedEvaporation();
        Penman _Penman = new Penman();
        PtSoil _PtSoil = new PtSoil();
        SoilEvaporation _SoilEvaporation = new SoilEvaporation();
        EvapoTranspiration _EvapoTranspiration = new EvapoTranspiration();
        SoilHeatFlux _SoilHeatFlux = new SoilHeatFlux();
        PotentialTranspiration _PotentialTranspiration = new PotentialTranspiration();
        CropHeatFlux _CropHeatFlux = new CropHeatFlux();
        CanopyTemperature _CanopyTemperature = new CanopyTemperature();

        private void EstimateOfAssociatedClasses(SiriusQualityEnergyBalance.DomainClass.EnergyBalanceState s,SiriusQualityEnergyBalance.DomainClass.EnergyBalanceState s1,SiriusQualityEnergyBalance.DomainClass.EnergyBalanceRate r,SiriusQualityEnergyBalance.DomainClass.EnergyBalanceAuxiliary a,SiriusQualityEnergyBalance.DomainClass.EnergyBalanceExogenous ex)
        {
            _NetRadiation.Estimate(s,s1, r, a, ex);
            _Conductance.Estimate(s,s1, r, a, ex);
            _DiffusionLimitedEvaporation.Estimate(s,s1, r, a, ex);
            _NetRadiationEquivalentEvaporation.Estimate(s,s1, r, a, ex);
            _PriestlyTaylor.Estimate(s,s1, r, a, ex);
            _PtSoil.Estimate(s,s1, r, a, ex);
            _Penman.Estimate(s,s1, r, a, ex);
            _SoilEvaporation.Estimate(s,s1, r, a, ex);
            _EvapoTranspiration.Estimate(s,s1, r, a, ex);
            _SoilHeatFlux.Estimate(s,s1, r, a, ex);
            _PotentialTranspiration.Estimate(s,s1, r, a, ex);
            _CropHeatFlux.Estimate(s,s1, r, a, ex);
            _CanopyTemperature.Estimate(s,s1, r, a, ex);
        }

        public EnergyBalanceComponent(EnergyBalanceComponent toCopy): this() // copy constructor 
        {
                albedoCoefficient = toCopy.albedoCoefficient;
                stefanBoltzman = toCopy.stefanBoltzman;
                elevation = toCopy.elevation;
                lambdaV = toCopy.lambdaV;
                psychrometricConstant = toCopy.psychrometricConstant;
                Alpha = toCopy.Alpha;
                vonKarman = toCopy.vonKarman;
                heightWeatherMeasurements = toCopy.heightWeatherMeasurements;
                zm = toCopy.zm;
                d = toCopy.d;
                zh = toCopy.zh;
                soilDiffusionConstant = toCopy.soilDiffusionConstant;
                rhoDensityAir = toCopy.rhoDensityAir;
                specificHeatCapacityAir = toCopy.specificHeatCapacityAir;
                tauAlpha = toCopy.tauAlpha;
                isWindVpDefined = toCopy.isWindVpDefined;
        }
    }
}

File: src/java/SQ_Energy_Balance/EnergyBalanceComponent.java
public class EnergyBalanceComponent
{
    
    public EnergyBalanceComponent() { }

    NetRadiation _NetRadiation = new NetRadiation();
    NetRadiationEquivalentEvaporation _NetRadiationEquivalentEvaporation = new NetRadiationEquivalentEvaporation();
    PriestlyTaylor _PriestlyTaylor = new PriestlyTaylor();
    Conductance _Conductance = new Conductance();
    DiffusionLimitedEvaporation _DiffusionLimitedEvaporation = new DiffusionLimitedEvaporation();
    Penman _Penman = new Penman();
    PtSoil _PtSoil = new PtSoil();
    SoilEvaporation _SoilEvaporation = new SoilEvaporation();
    EvapoTranspiration _EvapoTranspiration = new EvapoTranspiration();
    SoilHeatFlux _SoilHeatFlux = new SoilHeatFlux();
    PotentialTranspiration _PotentialTranspiration = new PotentialTranspiration();
    CropHeatFlux _CropHeatFlux = new CropHeatFlux();
    CanopyTemperature _CanopyTemperature = new CanopyTemperature();

    public Double getalbedoCoefficient()
    { return _NetRadiation.getalbedoCoefficient(); }
    public void setalbedoCoefficient(Double _albedoCoefficient){
    _NetRadiation.setalbedoCoefficient(_albedoCoefficient);
    }

    public Double getstefanBoltzman()
    { return _NetRadiation.getstefanBoltzman(); }
    public void setstefanBoltzman(Double _stefanBoltzman){
    _NetRadiation.setstefanBoltzman(_stefanBoltzman);
    }

    public Double getelevation()
    { return _NetRadiation.getelevation(); }
    public void setelevation(Double _elevation){
    _NetRadiation.setelevation(_elevation);
    }

    public Double getlambdaV()
    { return _NetRadiationEquivalentEvaporation.getlambdaV(); }
    public void setlambdaV(Double _lambdaV){
    _NetRadiationEquivalentEvaporation.setlambdaV(_lambdaV);
    _Penman.setlambdaV(_lambdaV);
    _CanopyTemperature.setlambdaV(_lambdaV);
    }

    public Double getpsychrometricConstant()
    { return _PriestlyTaylor.getpsychrometricConstant(); }
    public void setpsychrometricConstant(Double _psychrometricConstant){
    _PriestlyTaylor.setpsychrometricConstant(_psychrometricConstant);
    _Penman.setpsychrometricConstant(_psychrometricConstant);
    }

    public Double getAlpha()
    { return _PriestlyTaylor.getAlpha(); }
    public void setAlpha(Double _Alpha){
    _PriestlyTaylor.setAlpha(_Alpha);
    _Penman.setAlpha(_Alpha);
    _PtSoil.setAlpha(_Alpha);
    }

    public Double getvonKarman()
    { return _Conductance.getvonKarman(); }
    public void setvonKarman(Double _vonKarman){
    _Conductance.setvonKarman(_vonKarman);
    }

    public Double getheightWeatherMeasurements()
    { return _Conductance.getheightWeatherMeasurements(); }
    public void setheightWeatherMeasurements(Double _heightWeatherMeasurements){
    _Conductance.setheightWeatherMeasurements(_heightWeatherMeasurements);
    }

    public Double getzm()
    { return _Conductance.getzm(); }
    public void setzm(Double _zm){
    _Conductance.setzm(_zm);
    }

    public Double getd()
    { return _Conductance.getd(); }
    public void setd(Double _d){
    _Conductance.setd(_d);
    }

    public Double getzh()
    { return _Conductance.getzh(); }
    public void setzh(Double _zh){
    _Conductance.setzh(_zh);
    }

    public Double getsoilDiffusionConstant()
    { return _DiffusionLimitedEvaporation.getsoilDiffusionConstant(); }
    public void setsoilDiffusionConstant(Double _soilDiffusionConstant){
    _DiffusionLimitedEvaporation.setsoilDiffusionConstant(_soilDiffusionConstant);
    }

    public Double getrhoDensityAir()
    { return _Penman.getrhoDensityAir(); }
    public void setrhoDensityAir(Double _rhoDensityAir){
    _Penman.setrhoDensityAir(_rhoDensityAir);
    _CanopyTemperature.setrhoDensityAir(_rhoDensityAir);
    }

    public Double getspecificHeatCapacityAir()
    { return _Penman.getspecificHeatCapacityAir(); }
    public void setspecificHeatCapacityAir(Double _specificHeatCapacityAir){
    _Penman.setspecificHeatCapacityAir(_specificHeatCapacityAir);
    _CanopyTemperature.setspecificHeatCapacityAir(_specificHeatCapacityAir);
    }

    public Double gettau()
    { return _PtSoil.gettau(); }
    public void settau(Double _tau){
    _PtSoil.settau(_tau);
    _SoilHeatFlux.settau(_tau);
    _PotentialTranspiration.settau(_tau);
    }

    public Double gettauAlpha()
    { return _PtSoil.gettauAlpha(); }
    public void settauAlpha(Double _tauAlpha){
    _PtSoil.settauAlpha(_tauAlpha);
    }

    public Integer getisWindVpDefined()
    { return _EvapoTranspiration.getisWindVpDefined(); }
    public void setisWindVpDefined(Integer _isWindVpDefined){
    _EvapoTranspiration.setisWindVpDefined(_isWindVpDefined);
    }
    public void  Calculate_Model(EnergyBalanceState s, EnergyBalanceState s1, EnergyBalanceRate r, EnergyBalanceAuxiliary a, EnergyBalanceExogenous ex)
    {
        _NetRadiation.Calculate_Model(s, s1, r, a, ex);
        _Conductance.Calculate_Model(s, s1, r, a, ex);
        _DiffusionLimitedEvaporation.Calculate_Model(s, s1, r, a, ex);
        _NetRadiationEquivalentEvaporation.Calculate_Model(s, s1, r, a, ex);
        _PriestlyTaylor.Calculate_Model(s, s1, r, a, ex);
        _PtSoil.Calculate_Model(s, s1, r, a, ex);
        _Penman.Calculate_Model(s, s1, r, a, ex);
        _SoilEvaporation.Calculate_Model(s, s1, r, a, ex);
        _EvapoTranspiration.Calculate_Model(s, s1, r, a, ex);
        _SoilHeatFlux.Calculate_Model(s, s1, r, a, ex);
        _PotentialTranspiration.Calculate_Model(s, s1, r, a, ex);
        _CropHeatFlux.Calculate_Model(s, s1, r, a, ex);
        _CanopyTemperature.Calculate_Model(s, s1, r, a, ex);
    }
    private Double albedoCoefficient;
    private Double stefanBoltzman;
    private Double elevation;
    private Double lambdaV;
    private Double psychrometricConstant;
    private Double Alpha;
    private Double vonKarman;
    private Double heightWeatherMeasurements;
    private Double zm;
    private Double d;
    private Double zh;
    private Double soilDiffusionConstant;
    private Double rhoDensityAir;
    private Double specificHeatCapacityAir;
    private Double tau;
    private Double tauAlpha;
    private Integer isWindVpDefined;
    public EnergyBalanceComponent(EnergyBalanceComponent toCopy) // copy constructor 
    {
        this.albedoCoefficient = toCopy.getalbedoCoefficient();
        this.stefanBoltzman = toCopy.getstefanBoltzman();
        this.elevation = toCopy.getelevation();
        this.lambdaV = toCopy.getlambdaV();
        this.psychrometricConstant = toCopy.getpsychrometricConstant();
        this.Alpha = toCopy.getAlpha();
        this.vonKarman = toCopy.getvonKarman();
        this.heightWeatherMeasurements = toCopy.getheightWeatherMeasurements();
        this.zm = toCopy.getzm();
        this.d = toCopy.getd();
        this.zh = toCopy.getzh();
        this.soilDiffusionConstant = toCopy.getsoilDiffusionConstant();
        this.rhoDensityAir = toCopy.getrhoDensityAir();
        this.specificHeatCapacityAir = toCopy.getspecificHeatCapacityAir();
        this.tau = toCopy.gettau();
        this.tauAlpha = toCopy.gettauAlpha();
        this.isWindVpDefined = toCopy.getisWindVpDefined();

    }
}